pkg/resources/ray/deployment.go

```diff
@@ -0,0 +1,252 @@
+package ray
+
+import (
+	"fmt"
+	"strings"
+
+	appsv1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/intstr"
+	"k8s.io/utils/pointer"
+
+	dcv1alpha1 "github.com/dominodatalab/distributed-compute-operator/api/v1alpha1"
+	"github.com/dominodatalab/distributed-compute-operator/pkg/util"
+)
+
+var (
+	defaultCmd = []string{ApplicationName}
+	defaultEnv = []corev1.EnvVar{
+		{
+			Name: "MY_POD_IP",
+			ValueFrom: &corev1.EnvVarSource{
+				FieldRef: &corev1.ObjectFieldSelector{
+					FieldPath: "status.podIP",
+				},
+			},
+		},
+		{
+			Name: "MY_CPU_REQUEST",
+			ValueFrom: &corev1.EnvVarSource{
+				ResourceFieldRef: &corev1.ResourceFieldSelector{
+					Resource: "requests.cpu",
+				},
+			},
+		},
+	}
+	defaultVolumes = []corev1.Volume{
+		{
+			Name: sharedMemoryVolumeName,
+			VolumeSource: corev1.VolumeSource{
+				EmptyDir: &corev1.EmptyDirVolumeSource{
+					Medium: corev1.StorageMediumMemory,
+				},
+			},
+		},
+	}
+	defaultVolumeMounts = []corev1.VolumeMount{
+		{
+			Name:      sharedMemoryVolumeName,
+			MountPath: "/dev/shm",
+		},
+	}
+)
+
+const sharedMemoryVolumeName = "dshm"
+
+// NewDeployment generates a Deployment configured to manage Ray cluster nodes.
+// The configuration is based the provided spec and the desired Component workload.
+func NewDeployment(rc *dcv1alpha1.RayCluster, comp Component) (*appsv1.Deployment, error) {
+	var replicas int32
+	var nodeAttrs dcv1alpha1.RayClusterNode
+
+	switch comp {
+	case ComponentHead:
+		replicas = 1
+		nodeAttrs = rc.Spec.Head.RayClusterNode
+	case ComponentWorker:
+		replicas = *rc.Spec.Worker.Replicas
+		nodeAttrs = rc.Spec.Worker.RayClusterNode
+	default:
+		return nil, fmt.Errorf("invalid ray component: %q", comp)
+	}
+
+	imageRef, err := util.ParseImageDefinition(rc.Spec.Image)
+	if err != nil {
+		return nil, err
+	}
+
+	args := processArgs(rc, comp)
+	ports := processPorts(rc, comp)
+	labels := processLabels(rc, comp, nodeAttrs.Labels)
+	envVars := append(defaultEnv, rc.Spec.EnvVars...)
+	volumes := append(defaultVolumes, nodeAttrs.Volumes...)
+	volumeMounts := append(defaultVolumeMounts, nodeAttrs.VolumeMounts...)
+
+	serviceAccountName := InstanceObjectName(rc.Name, ComponentNone)
+	if rc.Spec.ServiceAccountName != "" {
+		serviceAccountName = rc.Spec.ServiceAccountName
+	}
+
+	deploy := &appsv1.Deployment{
+		TypeMeta: metav1.TypeMeta{
+			Kind:       "Deployment",
+			APIVersion: "apps/v1",
+		},
```

Comment:
This isn't interpreted from the type of the ... type?

Reply:
You're correct. Technically, adding type metadata is only REQUIRED when using the dynamic client to create objects from unstructured data. I added this to enrich my logs during `create` operations; `update` and `delete` already have the type metadata added by the API server.

```
# with type metadata
distributed-compute-operator 2021-03-09T19:49:59.965Z     INFO    controllers.RayCluster creating controlled object      {"raycluster": "default/example", "object": {"apiVersion": "v1", "kind": "ServiceAccount", "namespace": "default", "name": "example-ray"}}
distributed-compute-operator 2021-03-09T19:49:59.969Z     INFO    controllers.RayCluster creating controlled object      {"raycluster": "default/example", "object": {"apiVersion": "v1", "kind": "Service", "namespace": "default", "name": "example-ray-head"}}
distributed-compute-operator 2021-03-09T19:49:59.993Z     INFO    controllers.RayCluster creating controlled object      {"raycluster": "default/example", "object": {"apiVersion": "networking.k8s.io/v1", "kind": "NetworkPolicy", "namespace": "default", "name": "example-ray-cluster"}}
distributed-compute-operator 2021-03-09T19:49:59.998Z     INFO    controllers.RayCluster creating controlled object      {"raycluster": "default/example", "object": {"apiVersion": "networking.k8s.io/v1", "kind": "NetworkPolicy", "namespace": "default", "name": "example-ray-client"}}
distributed-compute-operator 2021-03-09T19:50:00.003Z     INFO    controllers.RayCluster creating controlled object      {"raycluster": "default/example", "object": {"apiVersion": "autoscaling/v2beta2", "kind": "HorizontalPodAutoscaler", "namespace": "default", "name": "example-ray"}}

# without type metadata
distributed-compute-operator 2021-03-09T19:49:59.965Z     INFO    controllers.RayCluster creating controlled object      {"raycluster": "default/example", "object": {"namespace": "default", "name": "example-ray"}}
distributed-compute-operator 2021-03-09T19:49:59.969Z     INFO    controllers.RayCluster creating controlled object      {"raycluster": "default/example", "object": {"namespace": "default", "name": "example-ray-head"}}
distributed-compute-operator 2021-03-09T19:49:59.993Z     INFO    controllers.RayCluster creating controlled object      {"raycluster": "default/example", "object": {"namespace": "default", "name": "example-ray-cluster"}}
distributed-compute-operator 2021-03-09T19:49:59.998Z     INFO    controllers.RayCluster creating controlled object      {"raycluster": "default/example", "object": {"namespace": "default", "name": "example-ray-client"}}
distributed-compute-operator 2021-03-09T19:50:00.003Z     INFO    controllers.RayCluster creating controlled object      {"raycluster": "default/example", "object": {"namespace": "default", "name": "example-ray"}}
```

I could remove this from the object generators and add it inside `RayClusterReconciler.createOrUpdateOwnedResource()` instead, but that's just another form of 💩. The current implementation is gross but I cannot think of any other way to accomplish this right now. You?

Reply:
if it's literally just for logging, does something like this work?
 ```
 		kinds, _, err := r.Scheme.ObjectKinds(controlled)
		if err != nil {
			return err
		}
		
		log.Info("creating controlled object", "kind", kinds[0], "object", controlled)
```

Reply:
It does work, awesome find. I'm still learning how to use `runtime.Scheme`. 🙃

Reply:
done.